libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,223 @@
+    // the same expiry the last added timer is later in the list.
+    s_soft_timer_array[timer_id].next_timer = s_soft_timer_array[last_node_id].next_timer;
+    s_soft_timer_array[last_node_id].next_timer = timer_id;
+    s_soft_timer_array[s_soft_timer_array[timer_id].next_timer].previous_timer = timer_id;
+  } else {
+    // Longer than the next node recurse to the next node.
+    prv_insert_timer(timer_id, s_soft_timer_array[last_node_id].next_timer);
+  }
+}
+
+void soft_timer_init(void) {
+  // Start the PeiphClock
+  RCC_APB1PeriphClockCmd(RCC_APB1Periph_TIM2, ENABLE);
+
+  // Stop the timer if it was running.
+  TIM_Cmd(TIM2, DISABLE);
+
+  // Get the clock speed of the clocks to determine the SYSCLK speed which is what TIM2 uses.
+  // Note this is susceptible to prescaling on both the AHB and APB but by default these are not
+  // scaled.
+  RCC_ClocksTypeDef clock_speeds;
+  RCC_GetClocksFreq(&clock_speeds);
+
+  // Configure each clock tick to be 1 microsecond from 0 to UINT32_MAX.
+  // Note that while this is valid for now, any prescale changes to AHB and APB will impact the
+  // Prescaler here.
+  TIM_TimeBaseInitTypeDef init_struct = { .TIM_Prescaler = clock_speeds.SYSCLK_Frequency / 1000000,
+                                          .TIM_CounterMode = TIM_CounterMode_Up,
+                                          .TIM_Period = UINT32_MAX,
+                                          .TIM_ClockDivision = TIM_CKD_DIV1 };
+  TIM_TimeBaseInit(TIM2, &init_struct);
+
+  // Reset the counter to 0.
+  TIM_SetCounter(TIM2, 0);
+
+  // Clear and disable all the timers and forget the last running timer and rollover count.
+  s_rollover_count = 0;
+  s_active_timer_id = SOFT_TIMER_MAX_TIMERS;
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    s_soft_timer_array[i].inuse = false;
+    s_soft_timer_array[i].next_timer = SOFT_TIMER_MAX_TIMERS;
+    s_soft_timer_array[i].previous_timer = SOFT_TIMER_MAX_TIMERS;
+  }
+
+  // Enable the interrupts on Capture Compare and updates. TIM2 uses IRQ channel 15.
+  stm32f0xx_interrupt_nvic_enable(15, INTERRUPT_PRIORITY_NORMAL);
+  TIM_ITConfig(TIM2, TIM_IT_CC1, ENABLE);
+  TIM_ClearITPendingBit(TIM2, TIM_IT_CC1);
+
+  // Update on overflows only. Clear any pending overflows.
+  TIM_UpdateRequestConfig(TIM2, TIM_UpdateSource_Regular);
+  TIM_ClearITPendingBit(TIM2, TIM_IT_Update);
+
+  // Start the timer.
+  TIM_Cmd(TIM2, ENABLE);
+  // Enable the overflow interrupt after to prevent an accidental overflow trigger.
+  TIM_ITConfig(TIM2, TIM_IT_Update, ENABLE);
+}
+
+StatusCode soft_timer_start(uint32_t duration_us, SoftTimerCallback callback, void *context,
+                            SoftTimerID *timer_id) {
+  uint32_t count = TIM_GetCounter(TIM2);
+
+  // Enable a critical section.
+  bool critical = critical_section_start();
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    if (!s_soft_timer_array[i].inuse) {
```

Comment:
One trick I use in the object pool is to use a bitset and the __builtin_ffsll intrinsic to get the next free node. That saves me from having to iterate over every element in the array.